### DISCUSSION TITLE
Failing to sample from Dirichlet Process Mixture Model using SGLD

### DISCUSSION THREAD

**Discussion by Habush** [blackjax-devs/blackjax#260] [Q&A] [answered]
### Bug Description

I'm trying to reproduce the example from tensorflow-probability titled [Fitting Dirichlet Process Mixture Model Using Preconditioned Stochastic Gradient Langevin Dynamics](https://www.tensorflow.org/probability/examples/Fitting_DPMM_Using_pSGLD#2_model) by using blackjax api. I've used the [SGLD](https://github.com/blackjax-devs/blackjax/blob/main/blackjax/sgmcmc/sgld.py) class. I tried to follow the example notebook that shows how to use SGMCMC in blackjax, but I can't seem to reproduce the results in the tensorflow example. 

### Steps/Code to Reproduce

I've implemented the blackjax version of the example in a Google Colab notebook located [here](https://colab.research.google.com/drive/1r7mN3ZKjDD6imlXaAUMMAHeVTG3qKR4J?usp=sharing)

### Expected Results

Samples from the posterior where there are three clusters

### Actual Results

I'm getting `nan` values for the variables I'm trying to sample.

### Versions

BlackJAX 0.8.2

### Additional comment

I suspect there are two causes of the issue:
     1. My definition of the log probability function
     2. the usage of the step size scheduling function

**Comment by rlouf** [comment]
Thank you for opening an issue. To help me debug, could you do the following:
1 - Look at the output of `logprior` and `loglikelihood` with test values for the parameters;
2 - Do the the same for the gradient estimator.
3 - Try with a constant schedule, using a `step_size` that is very, very small to begin with

**Comment by Habush** [comment]
Reducing the `step_size` from `1e-5` to `1e-12` and using a constant schedule, I now get real values instead of `nan`, so that made a change. However, it fails to converge to the target distribution as the number of inferred clusters is 19 instead of 3

**Comment by rlouf** [comment]
The step size is probably too small. Moving this to discussions as it does not seem to be a bug. Have you tried the parameters given in the TFP article?

```python
# Learning rates and decay
starter_learning_rate = 1e-6
end_learning_rate = 1e-10
decay_steps = 1e4

# Number of training steps
training_steps = 10000

# Mini-batch size
batch_size = 20

# Sample size for parameter posteriors
sample_size = 100
```

**Comment by junpenglao** [comment, marked as answer]
Most likely it is because you are trying to sample bounded variable, which results in non-valid proposal (log prob returns `nan`).
Recommend to check out https://github.com/blackjax-devs/blackjax/blob/main/examples/change_of_variable_hmc.ipynb

**Comment by Habush** [comment]
@junpenglao, you are right! I forgot about the constrained support of some of the distributions. After applying the required transformation, I'm able to get the same exact result as the example! (Thanks to @rlouf as well for taking the time to respond)

![image](https://user-images.githubusercontent.com/6312328/183510745-f64dc877-5f25-4779-a2ef-0736b6b8aba4.png)

![image](https://user-images.githubusercontent.com/6312328/183510790-c28e1848-40ce-4101-8b66-4c130b359433.png)

**Reply by rlouf** [reply]
That's awesome! Do you plan on making that example public?

**Comment by Habush** [comment]
If you are open to PR submissions, I can create an example notebook containing the blackjax version, and open a PR. Please let me know.

**Reply by rlouf** [reply]
By all means please submit a PR,

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

Heads-up for anyone landing here: the change-of-variable example has moved to the sampling-book — https://blackjax-devs.github.io/sampling-book/models/change_of_variable_hmc.html . The root cause in this thread was sampling bounded variables without a transform (log-prob returns `nan`); applying the change of variable fixes it, as confirmed above.
